toolchain/docs/check/pattern_matching.md

```diff
@@ -0,0 +1,285 @@
+   describes the pattern.
+2. Traverse the parse tree of the scrutinee expression to emit SemIR that
+   evaluates it.
+3. Traverse the pattern SemIR from step 1 (sometimes in conjunction with the
+   scrutinee SemIR) to emit SemIR that actually performs pattern matching.
+
+## Pattern insts
+
+The SemIR emitted in step 1 primarily consists of _pattern insts_, which are
+insts that describe the pattern itself. For example, given the pattern
+`(x: i32, y:i32)`, step 1 might emit the following SemIR:
+
+```
+%x.patt: %pattern_type.7ce = binding_pattern x [concrete]
+%y.patt: %pattern_type.7ce = binding_pattern y [concrete]
+%.loc4_21: %pattern_type.511 = tuple_pattern (%x.patt, %y.patt) [concrete]
+```
+
+Pattern insts do not represent executable code, and are generally ignored during
+lowering. Instead, they descriptively represent the pattern itself as a kind of
+constant value, and their primary consumer is step 3. The type of a pattern inst
+is a _pattern type_, which is represented by a `pattern_type` inst. For example,
+the `constants` block might define the types in the above SemIR like so:
+
+```
+%i32: type = class_type @Int, @Int(%int_32) [concrete]
+%pattern_type.7ce: type = pattern_type %i32 [concrete]
+%tuple.type: type = tuple_type (%i32, %i32) [concrete]
+%pattern_type.511: type = pattern_type %tuple.type [concrete]
+```
+
+We can read this as saying that the type of `%x.patt` and `%y.patt` is "pattern
+that matches an `i32` scrutinee", and the type of `%.loc4_21` is "pattern that
+matches a `(i32, i32)` scrutinee".
+
+Insts that are not pattern insts are called _procedural insts_. All the insts
+emitted in steps 2 and 3 are procedural insts, but step 1 can emit procedural
+insts as well. For example, in a pattern like `(x: i32, a + b)`, `i32` and
+`a + b` are ordinary expressions, and so their SemIR must be emitted during the
+initial traversal of the parse tree, as with any other expression.
+
+All the pattern insts for a given full-pattern are grouped together in a
+distinct block that contains no procedural insts. Consequently, `Check::Context`
+maintains `pattern_block_stack` as a separate `InstBlockStack` for pattern
+blocks, and provides separate methods like `AddPatternInst` for adding insts to
+it.
+
+## Inst ordering
+
+The SemIR produced in the first two steps is (like most SemIR) generally in
+bottom-up order, reflecting the order of the parse tree. However, the step 3
+traversal is performed top-down, starting with the root inst of the pattern and
+traversing into its dependencies.
+
+In some cases it is necessary for step 1 to allocate insts that won't actually
+be emitted until step 3, because they are responsible for performing pattern
+matching. When that happens, they are allocated but not added to a block, and
+their IDs are stored in the `Check::Context` (typically in a map from pattern
+inst IDs to the corresponding match inst IDs) so that they can be spliced into
```

Comment:
Is this specifically talking about the below maps? Is there some other use-case, for what's atypical? If this is it, maybe:

```suggestion
matching. When that happens, they are allocated but not added to a block, and
their IDs are stored in the `Check::Context` (in one of the two below maps from pattern
inst IDs to the corresponding match inst IDs) so that they can be spliced into
```

Reply:
I don't think there are any atypical examples right now, but see below regarding future-proofing.

Reply:
I'd really suggest "more may be added in the future" or similar instead of "typically", if you want more future-proofing. When you say "typically" and "for example", then I expect that there's _already_ an atypical case.

To try to explain how I think about this: the code may change, or it may not, it may change in some unpredictable way, or it can change alongside a doc update; trying to plan around what'll change in the future can be hard, unless you have a specific plan. However, if the documentation *already* says it's incomplete/unreliable, then it suggests people need to *already* read the code to understand what's going on. "Typically" in particular would make me hunt for what the "atypical" case is.

Reply:
That makes sense. How's this?